cluster/cluster.go

```diff
@@ -337,23 +344,55 @@ func (executor *GPDBExecutor) ExecuteLocalCommandWithContext(commandStr string,
 	return string(output), err
 }
 
+// Create a new exec.Command object so we can run it again
+func resetCmd(cmd *exec.Cmd) *exec.Cmd {
+	path := cmd.Path
+	args := cmd.Args
+	return exec.Command(path, args...)
+}
+
+/*
+ * ExecuteClusterCommandWithRetries, but only 1 attempt to keep the previous functionality
+ */
+func (executor *GPDBExecutor) ExecuteClusterCommand(scope Scope, commandList []ShellCommand) *RemoteOutput {
+	return executor.ExecuteClusterCommandWithRetries(scope, commandList, 1, 0)
+}
+
 /*
  * This function just executes all of the commands passed to it in parallel; it
  * doesn't care about the scope of the command except to pass that on to the
  * RemoteOutput after execution.
+ *
+ * It will retry the command up to maxAttempts times
  * TODO: Add batching to prevent bottlenecks when executing in a huge cluster.
  */
-func (executor *GPDBExecutor) ExecuteClusterCommand(scope Scope, commandList []ShellCommand) *RemoteOutput {
+func (executor *GPDBExecutor) ExecuteClusterCommandWithRetries(scope Scope, commandList []ShellCommand, maxAttempts int, retrySleep time.Duration) *RemoteOutput {
 	length := len(commandList)
 	finished := make(chan int)
 	numErrors := 0
 	for i := range commandList {
 		go func(index int) {
+			var (
+				out    []byte
+				err    error
+				stderr bytes.Buffer
+			)
 			command := commandList[index]
-			var stderr bytes.Buffer
-			cmd := command.Command
-			cmd.Stderr = &stderr
-			out, err := cmd.Output()
+			for attempt := 1; attempt <= maxAttempts; attempt++ {
+				stderr.Reset()
+				cmd := resetCmd(command.Command)
+				cmd.Stderr = &stderr
+				out, err = cmd.Output()
+				if err == nil {
+					break
+				} else {
+					newRetryErr := fmt.Errorf("attempt %d: error was %w: %s", attempt, err, stderr.String())
+					command.RetryError = joinerrs.Join(command.RetryError, newRetryErr)
+					if attempt != maxAttempts {
+						time.Sleep(retrySleep)
+					}
+				}
+			}
 			command.Stdout = string(out)
 			command.Stderr = stderr.String()
 			command.Error = err
```

Comment:
I want to double check my understanding of the new retry behavior and how it affects existing the `command.Error`.

Suppose there are two attempts for running a particular command. Attempt 1 fails so we retry on attempt 2. Attempt 2 succeeds so now `Command.Error` is `nil` but `command.RetryError` has the first error inside. If this is correct, should the number of errors, `numErrors`, be 0 or 1? I wondering if we should pass that info in the return of `NewRemoteOutput`.

Reply:
I wanted numErrors to only reflect the number of commands that were totally failed at the end